### PR TITLE
gerritVerifyStatusPush: gerrit reporter using gerrit verify-status plugin

### PR DESCRIPTION
**Pull request by tardyp**
Gerrit 2.13 has a new verify-status plugin which allows several CI status to be pushed for the same commit (like for github)

https://gerrit.googlesource.com/plugins/verify-status/

### CODE REVIEW

File: master/buildbot/reporters/gerrit_verify_status.py [hunk 1]
```diff
@@ -0,0 +1,198 @@
...
+            payload['reporter'] = reporter
+
+        if category is not None:
+            payload['category'] = category
+
+        if duration is not None:
+            payload['duration'] = duration
+
+        if self.verbose:
+            log.info(
```
Comment (sa2ajj) [inline master/buildbot/reporters/gerrit_verify_status.py]: maybe change this to `.debug` and remove the condition?

File: master/buildbot/reporters/gerrit_verify_status.py [hunk 2]
```diff
@@ -0,0 +1,198 @@
...
+        self._http = yield httpclientservice.HTTPClientService.getService(
+            self.master, baseURL, auth=auth)
+
+        self.verification_name = verification_name or Interpolate('%(prop:buildername)s')
+        self.reporter = reporter or "buildbot"
+        self.abstain = abstain
+        self.category = category
+        self.startDescription = startDescription or 'Build started.'
+        self.endDescription = endDescription or 'Build done.'
+        self.verbose = verbose
```
Comment (sa2ajj) [inline master/buildbot/reporters/gerrit_verify_status.py]: do we really need this?  with the fine grained control over the logging this seems to be unnecessary [any more]?

File: master/buildbot/reporters/gerrit_verify_status.py [hunk 3]
```diff
@@ -0,0 +1,198 @@
...
+                        verbose=False, **kwargs):
+        yield http.HttpStatusPushBase.reconfigService(self, **kwargs)
+
+        if baseURL.endswith('/'):
+            baseURL = baseURL[:-1]
+
+        self._http = yield httpclientservice.HTTPClientService.getService(
+            self.master, baseURL, auth=auth)
+
+        self.verification_name = verification_name or Interpolate('%(prop:buildername)s')
```
Comment (sa2ajj) [inline master/buildbot/reporters/gerrit_verify_status.py]: which of these data members we really want to expose as the part of the API?

File: master/buildbot/reporters/gerrit_verify_status.py [hunk 4]
```diff
@@ -0,0 +1,198 @@
...
+        :param url: The url link to more info about this job
+        :reporter: The user that verified this job
+        :category: A category for this job
+        "duration": The time it took to run this job
+
+        :return: A deferred with the result from Gerrit.
+        """
+        payload = {
+            'name': name,
+            'value': value}
```
Comment (sa2ajj) [inline master/buildbot/reporters/gerrit_verify_status.py]: [nit] i'd put `}` on a separate line

File: master/buildbot/reporters/gerrit_verify_status.py [hunk 5]
```diff
@@ -0,0 +1,198 @@
...
+
+    def formatDuration(self, duration):
+        """Format the duration.
+
+        This method could be overriden if really needed, as the duration format in gerrit
+        is an arbitrary string.
+        :param duration: duration in timedelta
+        """
+        days = duration.days
+        hours, remainder = divmod(duration.seconds, 3600)
```
Comment (sa2ajj) [inline master/buildbot/reporters/gerrit_verify_status.py]: [side note] i'm somewhat surprise `timedelta` does not provide these considering what its constructor looks like

File: master/buildbot/reporters/gerrit_verify_status.py [hunk 6]
```diff
@@ -0,0 +1,198 @@
...
+        if self.verbose:
+            log.info(
+                'Sending Gerrit status for {change_id}/{revision_id}: data={data}',
+                change_id=change_id, revision_id=revision_id, data=payload)
+
+        return self._http.post('/'.join([
+            'changes', str(change_id), 'revisions', str(revision_id), 'verify-status~verifications']),
+            json=payload)
+
+    def formatDuration(self, duration):
```
Comment (sa2ajj) [inline master/buildbot/reporters/gerrit_verify_status.py]: is it a part of the API?  (i think it's OK to override "private" methods as well)

File: master/buildbot/reporters/gerrit_verify_status.py [hunk 7]
```diff
@@ -0,0 +1,198 @@
...
+        if days:
+            return '{} day{} {}h {}m {}s'.format(
+                days, "s" if days > 1 else "",
+                hours, minutes, seconds)
+        elif hours:
+            return '{}h {}m {}s'.format(hours, minutes, seconds)
+        else:
+            return '{}m {}s'.format(minutes, seconds)
+
+    def getGerritChanges(self, props):
```
Comment (sa2ajj) [inline master/buildbot/reporters/gerrit_verify_status.py]: `@staticmethod`?

on a related note, is it a part of the API?

File: master/buildbot/reporters/gerrit_verify_status.py [hunk 8]
```diff
@@ -0,0 +1,198 @@
...
+        return []
+
+    @defer.inlineCallbacks
+    def send(self, build):
+        props = Properties.fromDict(build['properties'])
+        if build['complete']:
+            value = self.RESULTS_TABLE.get(build['results'], self.DEFAULT_RESULT)
+            comment = yield props.render(self.endDescription)
+            duration = self.formatDuration(build['complete_at'] - build['started_at'])
+
```
Comment (sa2ajj) [inline master/buildbot/reporters/gerrit_verify_status.py]: why do we need this empty line?

File: master/buildbot/reporters/gerrit_verify_status.py [hunk 9]
```diff
@@ -0,0 +1,198 @@
...
+        else:
+            value = 0
+            comment = yield props.render(self.startDescription)
+            duration = 'pending'
+
+        print build['results'], value
```
Comment (sa2ajj) [inline master/buildbot/reporters/gerrit_verify_status.py]: i think logging should be used here (if it's not a debug print, in the latter case -- remove)

File: master/buildbot/test/fake/httpclientservice.py [hunk 10]
```diff
@@ -57,6 +57,7 @@ class HTTPClientService(service.SharedService):
 
     getName from the fake and getName from the real must return the same values.
     """
+    quiet = False
```
Comment (sa2ajj) [inline master/buildbot/test/fake/httpclientservice.py]: I think this and related changes should be a separate commit: otherwise it's a bit difficult to see what was introduced for the new functionality and what was otherwise required *but* unrelated

File: master/buildbot/test/fake/httpclientservice.py [hunk 11]
```diff
@@ -108,8 +109,9 @@ def assertNoOutstanding(self):
                               "expected more http requests:\n {!r}".format(self._expected))
 
     def _doRequest(self, method, ep, params=None, data=None, json=None):
-        log.debug("{method} {ep} {params!r} <- {data!r}",
-                  method=method, ep=ep, params=params, data=data or json)
+        if not self.quiet:
```
Comment (sa2ajj) [inline master/buildbot/test/fake/httpclientservice.py]: why do we need this condition since below we use `.debug`, then it's up to the global logging configuration...

Comment (tardyp) [inline master/buildbot/test/fake/httpclientservice.py]: its for testing the .verbose method. You can test expectWasQuiet, because the fake is verbose.

File: master/buildbot/test/fake/httpclientservice.py (hunk 10 above)
Comment (tardyp) [inline master/buildbot/test/fake/httpclientservice.py]: its related, needed to create the test_not_verbose test

File: master/buildbot/reporters/gerrit_verify_status.py (hunk 9 above)
Comment (tardyp) [inline master/buildbot/reporters/gerrit_verify_status.py]: oups. debug print

File: master/buildbot/reporters/gerrit_verify_status.py (hunk 2 above)
Comment (tardyp) [inline master/buildbot/reporters/gerrit_verify_status.py]: Well, it all http reporters have the same.

I think the logging system is more for advanced users, and that being able to easily debug by adding verbose to the reporter makes more sense for beginner.
It is not so much of a burden to maintain.

File: master/buildbot/reporters/gerrit_verify_status.py (hunk 5 above)
Comment (tardyp) [inline master/buildbot/reporters/gerrit_verify_status.py]: apparently not, according to http://stackoverflow.com/questions/538666/python-format-timedelta-to-string, where I borrowed the code.

File: master/buildbot/reporters/gerrit_verify_status.py (hunk 7 above)
Comment (tardyp) [inline master/buildbot/reporters/gerrit_verify_status.py]: It is meant to be overriden if really needed (as per the docstring)
